src/h3lib/lib/polygon.c

```diff
@@ -139,71 +114,20 @@ bool polygonContainsPoint(const GeoPolygon* geoPolygon, const BBox* bboxes,
     return contains;
 }
 
-/**
- * Create a bounding box from a simple polygon defined as an array of vertices.
- * Known limitations:
- * - Does not support polygons with two adjacent points > 180 degrees of
- *   longitude apart. These will be interpreted as crossing the antimeridian.
- * - Does not currently support polygons containing a pole.
- * @param verts    Array of vertices
- * @param numVerts Number of vertices
- * @param bbox     Output bbox
- */
-void bboxFromVertices(const GeoCoord* verts, int numVerts, BBox* bbox) {
-    // Early exit if there are no vertices
-    if (numVerts == 0) {
-        bbox->north = 0;
-        bbox->south = 0;
-        bbox->east = 0;
-        bbox->west = 0;
-        return;
-    }
-    double lat;
-    double lon;
-
-    bbox->south = DBL_MAX;
-    bbox->west = DBL_MAX;
-    bbox->north = -1.0 * DBL_MAX;
-    bbox->east = -1.0 * DBL_MAX;
-    bool isTransmeridian = false;
-
-    for (int i = 0; i < numVerts; i++) {
-        lat = verts[i].lat;
-        lon = verts[i].lon;
-        if (lat < bbox->south) bbox->south = lat;
-        if (lon < bbox->west) bbox->west = lon;
-        if (lat > bbox->north) bbox->north = lat;
-        if (lon > bbox->east) bbox->east = lon;
-        // check for arcs > 180 degrees longitude, flagging as transmeridian
-        if (fabs(lon - verts[(i + 1) % numVerts].lon) > M_PI) {
-            isTransmeridian = true;
-        }
-    }
-    // Swap east and west if transmeridian
-    if (isTransmeridian) {
-        double tmp = bbox->east;
-        bbox->east = bbox->west;
-        bbox->west = tmp;
-    }
-}
+#undef TYPE
+#undef IS_EMPTY
+#undef INIT_ITERATION
+#undef ITERATE
 
-/**
- * Create a bounding box from a Geofence
- * @param geofence Input Geofence
- * @param bbox     Output bbox
- */
-void bboxFromGeofence(const Geofence* geofence, BBox* bbox) {
-    bboxFromVertices(geofence->verts, geofence->numVerts, bbox);
-}
+// Define macros used in polygon algos for LinkedGeoLoop
+#define TYPE LinkedGeoLoop
+#define INIT_ITERATION INIT_ITERATION_LINKED_LOOP
+#define ITERATE ITERATE_LINKED_LOOP
+#define IS_EMPTY IS_EMPTY_LINKED_LOOP
 
-/**
- * Create a bounding box from a GeoPolygon
- * @param polygon Input GeoPolygon
- * @param bboxes  Output bboxes, one for the outer loop and one for each hole
- */
-void bboxesFromGeoPolygon(const GeoPolygon* polygon, BBox* bboxes) {
-    bboxFromGeofence(&polygon->geofence, &bboxes[0]);
-    for (int i = 0; i < polygon->numHoles; i++) {
-        bboxFromGeofence(&polygon->holes[i], &bboxes[i + 1]);
-    }
-}
+#include "polygonAlgos.h"
```

Comment:
What do you think about having separate .c files for the two implementations, and including this header only once per compilation unit?

Reply:
That might be a lot cleaner. I can split this up.

Reply:
On second thought: I think this is a really good suggestion, but it's going to play hell with the two other PRs I have stacked on top of this. I will make this change in a future diff.

Reply:
I like this suggestion, but in terms of optimization, I think it might hurt you slightly, depending on where the algorithm is used, because compilers aren't great at link time optimization (LTO).

For example:

func.h:
```
#define FUNC(x) void x() {}
```
a.h:
```
void a();
```
a.c:
```
FUNC(a)
```
b.h:
```
void b();
```
b.c:
```
FUNC(b)
```
main.c:
```
int main(void) { a(); b(); }
```
It is very likely that above sample would generate code that looks like this pseudo-assembly:
```
main:
    call a
    call b
```

Whereas doing this without the a and b files could easily result in an empty main:
main.c
```
FUNC(a)
FUNC(b)
int main() { a(); b(); }
```

Reply:
Wouldn't this be the case just from placing these functions in `polygon.c`? They're called from `algos.c` anyways.

Reply:
Ya so that's why I said depending on how it's used. I guess it won't make much of a difference here.